### PR TITLE
backport: merge bitcoin#22840, #22937, #23446, #23522, #24026, #24104, #24167, #20744, partial bitcoin#23469, #24169 (replace boost::filesystem with std::filesystem)

### PR DESCRIPTION
**Pull request by kwvg**
## Additional Information

* Depends on https://github.com/dashpay/dash/pull/6085
* Depends on https://github.com/dashpay/dash/pull/6137
* Dependency for https://github.com/dashpay/dash/pull/6150

## Breaking Changes

None observed.

## Checklist:

- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas **(note: N/A)**
- [x] I have added or updated relevant unit/integration/functional/e2e tests
- [x] I have made corresponding changes to the documentation **(note: N/A)**
- [x] I have assigned this pull request to a milestone _(for repository code-owners and collaborators only)_

### CODE REVIEW

**Review by UdjinM6 — APPROVED** [review]
utACK d0887c356c6a9238ff5299193550de87e197ab85

File: src/wallet/wallet.cpp [hunk 1]
```diff
@@ -5232,7 +5232,7 @@ bool CWallet::AutoBackupWallet(const fs::path& wallet_path, bilingual_str& error
         }
         if(fs::exists(sourceFile)) {
             try {
-                fs::copy_file(sourceFile, backupFile);
+                fs::copy_file(sourceFile, backupFile, fs::copy_option::fail_if_exists);
```
[the remark this answers is not in the document]
Comment (kwvg) [reply src/wallet/wallet.cpp]: Because `fs::copy_file` is constrained [here](https://github.com/dashpay/dash/commit/834083fd9b64ace48057a977151fe2edee288ee5#diff-3492213c413b2d41f5e11a676db8f69ff54f71a2f5a14b2f61ee919d3a9f5122R95-R101), effectively excluding the variant that doesn't ask for `copy_options`. Without the change, there would be a compiler error.

```
wallet/wallet.cpp:5235:17: error: too few arguments to function call, expected 3, have 2; did you mean 'boost::filesystem::copy_file'?
                fs::copy_file(sourceFile, backupFile);
                ^~
                boost::filesystem::copy_file
/src/dash/depends/x86_64-pc-linux-gnu/include/boost/filesystem/operations.hpp:392:13: note: 'boost::filesystem::copy_file' declared here
inline bool copy_file(path const& from, path const& to)
            ^
1 error generated.
make[2]: *** [Makefile:12507: wallet/libbitcoin_wallet_a-wallet.o] Error 1
make[2]: Leaving directory '/src/dash/src'
make[1]: *** [Makefile:18776: all-recursive] Error 1
make[1]: Leaving directory '/src/dash/src'
make: *** [Makefile:804: all-recursive] Error 1
```

File: src/wallet/walletdb.cpp [hunk 2]
```diff
@@ -1135,15 +1135,15 @@ std::unique_ptr<WalletDatabase> MakeDatabase(const fs::path& path, const Databas
 #ifdef USE_SQLITE
         return MakeSQLiteDatabase(path, options, status, error);
 #endif
-        error = Untranslated(strprintf("Failed to open database path '%s'. Build does not support SQLite database format.", path.string()));
+        error = Untranslated(strprintf("Failed to open database path '%s'. Build does not support SQLite database format.", fs::PathToString(path)));
```
Comment (knst) [inline src/wallet/walletdb.cpp:1138]: should not it be `fs::quoted(fs::PathToString(path))` here?

Comment (kwvg) [reply src/wallet/walletdb.cpp:1138]: Even in one of the more recent backports ([bitcoin#20744](https://github.com/bitcoin/bitcoin/pull/20744)), it appears that using `fs::PathToString(path)` is the norm ([source](https://github.com/bitcoin/bitcoin/blob/07269321f38e46e9e02f16d7cd135ea90692638d/src/wallet/walletdb.cpp#L1166-L1180)). Looking at even the last iterations of Bitcoin Core v23 (the version these changes first appeared), it's much the same ([source](https://github.com/bitcoin/bitcoin/blob/v23-final/src/wallet/walletdb.cpp#L1166-L1180)).

File: src/qt/bitcoin.cpp [hunk 3]
```diff
@@ -758,8 +758,8 @@ int GuiMain(int argc, char* argv[])
     }
     // Validate/set custom css directory
     if (gArgs.IsArgSet("-custom-css-dir")) {
-        fs::path customDir = fs::path(gArgs.GetArg("-custom-css-dir", ""));
-        QString strCustomDir = QString::fromStdString(customDir.string());
+        fs::path customDir = fs::PathFromString(gArgs.GetArg("-custom-css-dir", ""));
+        QString strCustomDir = QString::fromStdString(fs::PathToString(customDir));
```
Comment (knst) [inline src/qt/bitcoin.cpp:762]: Should be `PathToQString`` used instead of `QString::fromStdString(fs::PathToString))` ?

File: src/fs.cpp [hunk 4]
```diff
@@ -128,7 +128,7 @@ bool FileLock::TryLock()
     if (hFile == INVALID_HANDLE_VALUE) {
         return false;
     }
-    _OVERLAPPED overlapped = {};
+    _OVERLAPPED overlapped = {0};
```
Comment (knst) [inline src/fs.cpp:131]: 20744 - why that is added? it seems unrelated; moreover useless: there's no UB without `0`, it's completely fine

As I checked, bitcoin doesn't have it

File: src/util/system.cpp [hunk 5]
```diff
@@ -74,9 +74,14 @@
 #include <malloc.h>
 #endif
 
+#include <univalue.h>
```
Comment (knst) [inline src/util/system.cpp:77]: 20744 - partial due to missing changes in `src/util/syscall_sandbox.cpp` ?

File: src/fs.cpp (hunk 4 above)
Comment (kwvg) [reply src/fs.cpp]: Resolved in latest push!

File: src/interfaces/wallet.h [hunk 6]
```diff
@@ -347,7 +347,7 @@ class WalletLoader : public ChainClient
    virtual std::string getWalletDir() = 0;
 
    //! Restore backup wallet
-   virtual std::unique_ptr<Wallet> restoreWallet(const std::string& backup_file, const std::string& wallet_name, bilingual_str& error, std::vector<bilingual_str>& warnings) = 0;
+   virtual std::unique_ptr<Wallet> restoreWallet(const fs::path& backup_file, const std::string& wallet_name, bilingual_str& error, std::vector<bilingual_str>& warnings) = 0;
```
[the remark this answers is not in the document]
Comment (kwvg) [reply src/interfaces/wallet.h:350]: Resolved in latest push!

File: doc/build-unix.md [hunk 7]
```diff
@@ -45,11 +45,11 @@ Optional dependencies:
...
  libnatpmp   | NAT-PMP Support  | Firewall-jumping support
- libdb4.8    | Berkeley DB      | Optional, wallet storage (only needed when legacy wallet enabled)
+ libdb4.8    | Berkeley DB      | Wallet storage (only needed when legacy wallet enabled)
  qt          | GUI              | GUI toolkit (only needed when GUI enabled)
- libqrencode | QR codes in GUI  | Optional for generating QR codes (only needed when GUI enabled)
- libzmq3     | ZMQ notification | Optional, allows generating ZMQ notifications (requires ZMQ version >= 4.0.0)
- sqlite3     | SQLite DB        | Optional, wallet storage (only needed when descriptor wallet enabled)
+ libqrencode | QR codes in GUI  | QR code generation (only needed when GUI enabled)
+ libzmq3     | ZMQ notification | ZMQ notifications (requires ZMQ version >= 4.0.0)
+ sqlite3     | SQLite DB        | Wallet storage (only needed when descriptor wallet enabled)
```
[the remark this answers is not in the document]
Comment (kwvg) [reply doc/build-unix.md:52]: Resolved in latest push!

File: src/qt/rpcconsole.cpp [hunk 8]
```diff
@@ -814,7 +814,7 @@ void RPCConsole::addWallet(WalletModel * const walletModel)
         // The only loaded wallet
         ui->btn_rescan1->setEnabled(true);
         ui->btn_rescan2->setEnabled(true);
```
[the remark this answers is not in the document]
Comment (kwvg) [reply src/qt/rpcconsole.cpp:816]: Resolved in latest push!

File: src/qt/bitcoin.cpp (hunk 3 above)
Comment (kwvg) [reply src/qt/bitcoin.cpp]: Resolved in latest push!

File: src/util/system.cpp (hunk 5 above)
Comment (kwvg) [reply src/util/system.cpp:77]: Bitcoin eventually does away with syscall sandboxing entirely (see [bitcoin#27896](https://github.com/bitcoin/bitcoin/pull/27896)), it's very likely any backports related to it will be marked DNM.

**Review by knst — APPROVED** [review]
utACK  0f239203a890494b3f147c9060f039d7c2eb5c37

```
git range-diff 0f239203a890494b3f147c9060f039d7c2eb5c37~11..0f239203a890494b3f147c9060f039d7c2eb5c37 e754fa21b9764adb75a1027680027dfbc57897c4~11..e754fa21b9764adb75a1027680027dfbc57897c4
:
```
c4~11..e754fa21b9764adb75a1027680027dfbc57897c4
 1:  28b96a071d =  1:  4a4627f839 merge bitcoin#22840: fix unoptimized libraries in depends
 2:  23fe7e2f07 =  2:  41793c307c chore: dashify symbols in some unit tests
 3:  ecfac10b8e =  3:  f0194be1fc merge bitcoin#22937: Forbid calling unsafe fs::path(std::string) constructor and fs::path::string() method
 4:  193f6fde2e =  4:  9bf02baf8a merge bitcoin#23446: Mention that BerkeleyDB is for legacy wallet in build-unix
 5:  20d359b570 =  5:  9a678e837f partial bitcoin#23469: Remove Boost build note from build-unix.md
 6:  b0d2484a0b =  6:  5f53ba8c69 merge bitcoin#23522: Improve fs::PathToString documentation
 7:  7c270e6883 =  7:  bca606da6b merge bitcoin#24026: Block unsafe std::string fs::path conversion copy_file calls
 8:  7ffea4348f =  8:  02a3fcb7ca merge bitcoin#24104: Make compatible with boost 1.78
 9:  be7ac493d0 =  9:  25ece7e2bd merge bitcoin#24167: consistently use fsbridge:: for ifstream / ofstream
10:  a3b79267e0 = 10:  15d418e00b merge bitcoin#20744: Use std::filesystem. Remove Boost Filesystem & System
11:  0f239203a8 = 11:  e754fa21b9 partial bitcoin#24169: Add --enable-c++20 option
```

**Review by PastaPastaPasta — APPROVED** [review]
utACK 0f239203a890494b3f147c9060f039d7c2eb5c37
